### PR TITLE
Kotlinx Serialization: with a vengeance.

### PR DESCRIPTION
**Pull request by dbrant**
This does all of the following:

* Remove the actual dependency on Gson, for real.
* Remove the usage of `@Contextual` annotations, and instead use proper polymorphism specifiers.
* Create serializers for third-party classes (`Date`, `Location`, etc)
* Create a generic facility for serializing/deserializing arbitrary classes, lists, maps, etc. (`JsonUtil`)
* Update and/or fix all tests.
* Code reduction of over 1000 lines.

### CODE REVIEW

**Review by cooltey — CHANGES_REQUESTED** [review]
Found the following issues:
1. `kotlinx.serialization.SerializationException: Serializer for class 'SessionData' is not found.`
2. `kotlinx.serialization.json.internal.JsonDecodingException: Unexpected JSON token at offset 25568: Expected string literal but 'null' literal was found.
    Use 'coerceInputValues = true' in 'Json {}' builder to coerce nulls to default values.` and it causes error when opening the `NotifficationActivity`.
3. `kotlinx.serialization.SerializationException: Serializer for class 'SiteInfo' is not found` in `LangaugesListActivity`.
4. `Class not found when unmarshalling: org.wikipedia.userprofile.Contribution` in `ContributionsDetailsActivity`
5. `Class not found when unmarshalling: org.wikipedia.page.PageTitle` when opening the `EditSectionActivity`.
6. `kotlinx.serialization.json.internal.JsonDecodingException: Unexpected JSON token at offset 1282: Unexpected symbol 'n' in numeric literal
    JSON input: .....text": "", "offset": {"from": null,"to": 1965}},{"type": 1, .....` when opening Watchlist `ArticleEditDetailsFragment`

**Comment by dbrant** [comment]
@cooltey  Great catches:

> 1. `kotlinx.serialization.SerializationException: Serializer for class 'SessionData' is not found.`

Done.

> 2. `kotlinx.serialization.json.internal.JsonDecodingException: Unexpected JSON token at offset 25568: Expected string literal but 'null' literal was found. Use 'coerceInputValues = true' in 'Json {}' builder to coerce nulls to default values.` and it causes error when opening the `NotifficationActivity`.

Can't reproduce -- it must be dependent on the exact notifications that you have for you account. Can you paste more info from the exception that you get?

> 3. `kotlinx.serialization.SerializationException: Serializer for class 'SiteInfo' is not found` in `LangaugesListActivity`.

Done.

> 4. `Class not found when unmarshalling: org.wikipedia.userprofile.Contribution` in `ContributionsDetailsActivity`

Can't reproduce. Can you post the full exception you see?

> 5. `Class not found when unmarshalling: org.wikipedia.page.PageTitle` when opening the `EditSectionActivity`.

Can't reproduce. Can you post the full actual exception you see?

> 6. `kotlinx.serialization.json.internal.JsonDecodingException: Unexpected JSON token at offset 1282: Unexpected symbol 'n' in numeric literal JSON input: .....text": "", "offset": {"from": null,"to": 1965}},{"type": 1, .....` when opening Watchlist `ArticleEditDetailsFragment`

Done.

**Comment by cooltey** [comment]
@dbrant 
> > 1. `kotlinx.serialization.json.internal.JsonDecodingException: Unexpected JSON token at offset 25568: Expected string literal but 'null' literal was found. Use 'coerceInputValues = true' in 'Json {}' builder to coerce nulls to default values.` and it causes error when opening the `NotifficationActivity`.
> 
> Can't reproduce -- it must be dependent on the exact notifications that you have for your account. Can you paste more info from the exception that you get?

```
2021-09-23 14:48:26.512 7251-7251/org.wikipedia.dev E/org.wikipedia.notifications.NotificationActivity: setErrorState():203: 
    kotlinx.serialization.json.internal.JsonDecodingException: Unexpected JSON token at offset 25568: Expected string literal but 'null' literal was found.
    Use 'coerceInputValues = true' in 'Json {}` builder to coerce nulls to default values.
    JSON input: .....在「‪太空部隊‬」的連結通知","description":null,"data":{"tokenType":"csrf.....
        at kotlinx.serialization.json.internal.JsonExceptionsKt.JsonDecodingException(JsonExceptions.kt:24)
        at kotlinx.serialization.json.internal.JsonExceptionsKt.JsonDecodingException(JsonExceptions.kt:32)
        at kotlinx.serialization.json.internal.JsonLexer.fail(JsonLexer.kt:479)
        at kotlinx.serialization.json.internal.JsonLexer.unexpectedToken(JsonLexer.kt:196)
        at kotlinx.serialization.json.internal.JsonLexer.consumeNextToken(JsonLexer.kt:188)
        at kotlinx.serialization.json.internal.JsonLexer.consumeKeyString(JsonLexer.kt:298)
        at kotlinx.serialization.json.internal.JsonLexer.consumeString(JsonLexer.kt:318)
        at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeString(StreamingJsonDecoder.kt:245)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeStringElement(AbstractDecoder.kt:58)
        at org.wikipedia.notifications.Notification$Link$$serializer.deserialize(Notification.kt:72)
        at org.wikipedia.notifications.Notification$Link$$serializer.deserialize(Notification.kt:72)
        at kotlinx.serialization.json.internal.PolymorphicKt.decodeSerializableValuePolymorphic(Polymorphic.kt:63)
        at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:32)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:43)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableElement(AbstractDecoder.kt:70)
        at kotlinx.serialization.encoding.CompositeDecoder$DefaultImpls.decodeSerializableElement$default(Decoding.kt:535)
        at kotlinx.serialization.internal.ListLikeSerializer.readElement(CollectionSerializers.kt:80)
        at kotlinx.serialization.internal.AbstractCollectionSerializer.readElement$default(CollectionSerializers.kt:51)
        at kotlinx.serialization.internal.AbstractCollectionSerializer.merge(CollectionSerializers.kt:36)
        at kotlinx.serialization.internal.AbstractCollectionSerializer.deserialize(CollectionSerializers.kt:43)
        at kotlinx.serialization.json.internal.PolymorphicKt.decodeSerializableValuePolymorphic(Polymorphic.kt:63)
        at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:32)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:43)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeNullableSerializableElement(AbstractDecoder.kt:79)
        at org.wikipedia.notifications.Notification$Links$$serializer.deserialize(Notification.kt:83)
        at org.wikipedia.notifications.Notification$Links$$serializer.deserialize(Notification.kt:83)
        at kotlinx.serialization.json.internal.PolymorphicKt.decodeSerializableValuePolymorphic(Polymorphic.kt:63)
        at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:32)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:43)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeNullableSerializableElement(AbstractDecoder.kt:79)
        at org.wikipedia.notifications.Notification$Contents$$serializer.deserialize(Notification.kt:110)
        at org.wikipedia.notifications.Notification$Contents$$serializer.deserialize(Notification.kt:110)
        at kotlinx.serialization.json.internal.PolymorphicKt.decodeSerializableValuePolymorphic(Polymorphic.kt:63)
        at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:32)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:43)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeNullableSerializableElement(AbstractDecoder.kt:79)
2021-09-23 14:48:26.513 7251-7251/org.wikipedia.dev E/org.wikipedia.notifications.NotificationActivity:     at org.wikipedia.notifications.Notification$$serializer.deserialize(Notification.kt:11)
        at org.wikipedia.notifications.Notification$$serializer.deserialize(Notification.kt:11)
        at kotlinx.serialization.json.internal.PolymorphicKt.decodeSerializableValuePolymorphic(Polymorphic.kt:63)
        at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:32)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:43)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableElement(AbstractDecoder.kt:70)
        at kotlinx.serialization.encoding.CompositeDecoder$DefaultImpls.decodeSerializableElement$default(Decoding.kt:535)
        at kotlinx.serialization.internal.ListLikeSerializer.readElement(CollectionSerializers.kt:80)
        at kotlinx.serialization.internal.AbstractCollectionSerializer.readElement$default(CollectionSerializers.kt:51)
        at kotlinx.serialization.internal.AbstractCollectionSerializer.merge(CollectionSerializers.kt:36)
        at kotlinx.serialization.internal.AbstractCollectionSerializer.deserialize(CollectionSerializers.kt:43)
        at kotlinx.serialization.json.internal.PolymorphicKt.decodeSerializableValuePolymorphic(Polymorphic.kt:63)
        at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:32)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:43)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeNullableSerializableElement(AbstractDecoder.kt:79)
        at org.wikipedia.dataclient.mwapi.MwQueryResult$NotificationList$$serializer.deserialize(MwQueryResult.kt:144)
        at org.wikipedia.dataclient.mwapi.MwQueryResult$NotificationList$$serializer.deserialize(MwQueryResult.kt:144)
        at kotlinx.serialization.json.internal.PolymorphicKt.decodeSerializableValuePolymorphic(Polymorphic.kt:63)
        at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:32)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:43)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeNullableSerializableElement(AbstractDecoder.kt:79)
        at org.wikipedia.dataclient.mwapi.MwQueryResult$$serializer.deserialize(MwQueryResult.kt:14)
        at org.wikipedia.dataclient.mwapi.MwQueryResult$$serializer.deserialize(MwQueryResult.kt:14)
        at kotlinx.serialization.json.internal.PolymorphicKt.decodeSerializableValuePolymorphic(Polymorphic.kt:63)
        at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:32)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:43)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeNullableSerializableElement(AbstractDecoder.kt:79)
        at org.wikipedia.dataclient.mwapi.MwQueryResponse$$serializer.deserialize(MwQueryResponse.kt:6)
        at org.wikipedia.dataclient.mwapi.MwQueryResponse$$serializer.deserialize(MwQueryResponse.kt:6)
        at kotlinx.serialization.json.internal.PolymorphicKt.decodeSerializableValuePolymorphic(Polymorphic.kt:63)
        at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:32)
        at kotlinx.serialization.json.Json.decodeFromString(Json.kt:95)
        at com.jakewharton.retrofit2.converter.kotlinx.serialization.Serializer$FromString.fromResponseBody(Serializer.kt:30)
        at com.jakewharton.retrofit2.converter.kotlinx.serialization.DeserializationStrategyConverter.convert(DeserializationStrategyConverter.kt:11)
        at com.jakewharton.retrofit2.converter.kotlinx.serialization.DeserializationStrategyConverter.convert(DeserializationStrategyConverter.kt:7)
        at retrofit2.OkHttpCall.parseResponse(OkHttpCall.java:243)
        at retrofit2.OkHttpCall$1.onResponse(OkHttpCall.java:153)
        at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:519)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
2021-09-23 14:48:26.513 7251-7251/org.wikipedia.dev E/org.wikipedia.notifications.NotificationActivity:     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:919)

```
> 
> > 1. `Class not found when unmarshalling: org.wikipedia.userprofile.Contribution` in `ContributionsDetailsActivity`
> 
> Can't reproduce. Can you post the full exception you see?
> 
```
2021-09-23 14:49:54.771 7251-7251/org.wikipedia.dev E/org.wikipedia.diff.ArticleEditDetailsFragment: setErrorState():138: 
    kotlinx.serialization.json.internal.JsonDecodingException: Unexpected JSON token at offset 1282: Unexpected symbol 'n' in numeric literal
    JSON input: .....text": "", "offset": {"from": null,"to": 1965}},{"type": 1, .....
        at kotlinx.serialization.json.internal.JsonExceptionsKt.JsonDecodingException(JsonExceptions.kt:24)
        at kotlinx.serialization.json.internal.JsonExceptionsKt.JsonDecodingException(JsonExceptions.kt:32)
        at kotlinx.serialization.json.internal.JsonLexer.fail(JsonLexer.kt:479)
        at kotlinx.serialization.json.internal.JsonLexer.fail$default(JsonLexer.kt:478)
        at kotlinx.serialization.json.internal.JsonLexer.consumeNumericLiteral(JsonLexer.kt:517)
        at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeInt(StreamingJsonDecoder.kt:203)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeIntElement(AbstractDecoder.kt:53)
        at org.wikipedia.dataclient.restbase.DiffResponse$DiffOffset$$serializer.deserialize(DiffResponse.kt:24)
        at org.wikipedia.dataclient.restbase.DiffResponse$DiffOffset$$serializer.deserialize(DiffResponse.kt:24)
        at kotlinx.serialization.json.internal.PolymorphicKt.decodeSerializableValuePolymorphic(Polymorphic.kt:63)
        at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:32)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:43)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeNullableSerializableElement(AbstractDecoder.kt:79)
        at org.wikipedia.dataclient.restbase.DiffResponse$DiffItem$$serializer.deserialize(DiffResponse.kt:14)
        at org.wikipedia.dataclient.restbase.DiffResponse$DiffItem$$serializer.deserialize(DiffResponse.kt:14)
        at kotlinx.serialization.json.internal.PolymorphicKt.decodeSerializableValuePolymorphic(Polymorphic.kt:63)
        at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:32)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:43)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableElement(AbstractDecoder.kt:70)
        at kotlinx.serialization.encoding.CompositeDecoder$DefaultImpls.decodeSerializableElement$default(Decoding.kt:535)
        at kotlinx.serialization.internal.ListLikeSerializer.readElement(CollectionSerializers.kt:80)
        at kotlinx.serialization.internal.AbstractCollectionSerializer.readElement$default(CollectionSerializers.kt:51)
        at kotlinx.serialization.internal.AbstractCollectionSerializer.merge(CollectionSerializers.kt:36)
        at kotlinx.serialization.internal.AbstractCollectionSerializer.deserialize(CollectionSerializers.kt:43)
        at kotlinx.serialization.json.internal.PolymorphicKt.decodeSerializableValuePolymorphic(Polymorphic.kt:63)
        at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:32)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:43)
        at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableElement(AbstractDecoder.kt:70)
        at org.wikipedia.dataclient.restbase.DiffResponse$$serializer.deserialize(DiffResponse.kt:6)
        at org.wikipedia.dataclient.restbase.DiffResponse$$serializer.deserialize(DiffResponse.kt:6)
        at kotlinx.serialization.json.internal.PolymorphicKt.decodeSerializableValuePolymorphic(Polymorphic.kt:63)
        at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:32)
        at kotlinx.serialization.json.Json.decodeFromString(Json.kt:95)
        at com.jakewharton.retrofit2.converter.kotlinx.serialization.Serializer$FromString.fromResponseBody(Serializer.kt:30)
        at com.jakewharton.retrofit2.converter.kotlinx.serialization.DeserializationStrategyConverter.convert(DeserializationStrategyConverter.kt:11)
        at com.jakewharton.retrofit2.converter.kotlinx.serialization.DeserializationStrategyConverter.convert(DeserializationStrategyConverter.kt:7)

```

> > 1. `Class not found when unmarshalling: org.wikipedia.page.PageTitle` when opening the `EditSectionActivity`.
> 
> Can't reproduce. Can you post the full exception you see?
> 

```
2021-09-23 14:50:54.801 1387-11216/? E/Parcel: Class not found when unmarshalling: org.wikipedia.page.PageTitle
    java.lang.ClassNotFoundException: org.wikipedia.page.PageTitle
        at java.lang.Class.classForName(Native Method)
        at java.lang.Class.forName(Class.java:454)
        at android.os.Parcel.readParcelableCreator(Parcel.java:3031)
        at android.os.Parcel.readParcelable(Parcel.java:2981)
        at android.os.Parcel.readValue(Parcel.java:2883)
        at android.os.Parcel.readArrayMapInternal(Parcel.java:3261)
        at android.os.BaseBundle.initializeFromParcelLocked(BaseBundle.java:292)
        at android.os.BaseBundle.unparcel(BaseBundle.java:236)
        at android.os.BaseBundle.getString(BaseBundle.java:1160)
        at android.content.Intent.getStringExtra(Intent.java:8552)
        at com.android.server.wm.ActivityStarter.startActivity(ActivityStarter.java:847)
        at com.android.server.wm.ActivityStarter.startActivity(ActivityStarter.java:732)
        at com.android.server.wm.ActivityStarter.startActivityMayWait(ActivityStarter.java:2074)
        at com.android.server.wm.ActivityStarter.execute(ActivityStarter.java:646)
        at com.android.server.wm.ActivityTaskManagerService.startActivityAsUser(ActivityTaskManagerService.java:1717)
        at com.android.server.wm.ActivityTaskManagerService.startActivityAsUser(ActivityTaskManagerService.java:1613)
        at com.android.server.wm.ActivityTaskManagerService.startActivity(ActivityTaskManagerService.java:1566)
        at android.app.IActivityTaskManager$Stub.onTransact(IActivityTaskManager.java:1746)
        at android.os.Binder.execTransactInternal(Binder.java:1056)
        at android.os.Binder.execTransact(Binder.java:1029)
     Caused by: java.lang.ClassNotFoundException: org.wikipedia.page.PageTitle
        at java.lang.Class.classForName(Native Method) 
        at java.lang.Class.forName(Class.java:454) 
        at android.os.Parcel.readParcelableCreator(Parcel.java:3031) 
        at android.os.Parcel.readParcelable(Parcel.java:2981) 
        at android.os.Parcel.readValue(Parcel.java:2883) 
        at android.os.Parcel.readArrayMapInternal(Parcel.java:3261) 
        at android.os.BaseBundle.initializeFromParcelLocked(BaseBundle.java:292) 
        at android.os.BaseBundle.unparcel(BaseBundle.java:236) 
        at android.os.BaseBundle.getString(BaseBundle.java:1160) 
        at android.content.Intent.getStringExtra(Intent.java:8552) 
        at com.android.server.wm.ActivityStarter.startActivity(ActivityStarter.java:847) 
        at com.android.server.wm.ActivityStarter.startActivity(ActivityStarter.java:732) 
        at com.android.server.wm.ActivityStarter.startActivityMayWait(ActivityStarter.java:2074) 
        at com.android.server.wm.ActivityStarter.execute(ActivityStarter.java:646) 
        at com.android.server.wm.ActivityTaskManagerService.startActivityAsUser(ActivityTaskManagerService.java:1717) 
        at com.android.server.wm.ActivityTaskManagerService.startActivityAsUser(ActivityTaskManagerService.java:1613) 
        at com.android.server.wm.ActivityTaskManagerService.startActivity(ActivityTaskManagerService.java:1566) 
        at android.app.IActivityTaskManager$Stub.onTransact(IActivityTaskManager.java:1746) 
        at android.os.Binder.execTransactInternal(Binder.java:1056) 
        at android.os.Binder.execTransact(Binder.java:1029) 

```

**Comment by dbrant** [comment]
@cooltey  Thanks! All of these issues should now be fixed, except the errors that you see about `Class not found when unmarshalling`.  Can you try doing a `Clean Project` and rebuilding, and see if you still have problems?

**Comment by cooltey** [comment]
> 
> 
> @cooltey Thanks! All of these issues should now be fixed, except the errors that you see about `Class not found when unmarshalling`. Can you try doing a `Clean Project` and rebuilding, and see if you still have problems?

It works for me. I ran `Clean Project` twice and not seeing the error messages anymore.

File: app/src/main/java/org/wikipedia/json/DateSerializer.kt [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import kotlinx.serialization.descriptors.*
+import kotlinx.serialization.encoding.*
+import org.wikipedia.util.log.L
+import java.lang.Exception
+import java.util.*
+
+@Deprecated("This serializer is used only for HistoryEntry items in our Tabs structure and" +
+        "should be removed once we switch to saving tabs in the database instead of serializing.")
+object DateSerializer : KSerializer<Date> {
+    override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("Date", PrimitiveKind.LONG)
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/json/DateSerializer.kt:13]: Explicit type specification can be removed.

File: app/src/main/java/org/wikipedia/json/LocationSerializer.kt [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+import android.location.Location
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.descriptors.buildClassSerialDescriptor
+import kotlinx.serialization.descriptors.element
+import kotlinx.serialization.encoding.*
+
+object LocationSerializer : KSerializer<Location> {
+    override val descriptor: SerialDescriptor = buildClassSerialDescriptor("Location") {
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/json/LocationSerializer.kt:11]: Explicit type specification can be removed.

File: app/src/main/java/org/wikipedia/json/UriSerializer.kt [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package org.wikipedia.json
+
+import android.net.Uri
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.descriptors.*
+import kotlinx.serialization.encoding.*
+
+object UriSerializer : KSerializer<Uri> {
+    override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("Uri", PrimitiveKind.STRING)
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/json/UriSerializer.kt:9]: Explicit type specification can be removed.

**Review by cooltey — CHANGES_REQUESTED** [review]

**Comment by sharvaniharan** [comment]
Jsondecoding exceptioin while fetching stream configs
Screen Shot 2021-09-26 at 6 03 14 PM

**Comment by sharvaniharan** [comment]
I am also seeing that when we click on the lead image in an article, the license info doesnt load and progressbar is active infinitely...

File: app/src/main/java/org/wikipedia/edit/preview/EditPreviewFragment.kt [hunk 4]
```diff
@@ -184,10 +184,11 @@ class EditPreviewFragment : Fragment(), CommunicationBridgeListener, ReferenceDi
         bridge.addListener("media") { _, _ -> }
 
         bridge.addListener("reference") { _, messagePayload ->
-            references =
-                GsonUtil.getDefaultGson().fromJson(messagePayload, PageReferences::class.java)
-            if (references.referencesGroup!!.isNotEmpty()) {
-                bottomSheetPresenter.show(childFragmentManager, ReferenceDialog())
+            (JsonUtil.decodeFromString(messagePayload.toString()) as PageReferences?)?.let {
```
[the remark this answers is not in the document]
Comment (dbrant) [reply app/src/main/java/org/wikipedia/edit/preview/EditPreviewFragment.kt:187]: yes!

File: app/src/main/java/org/wikipedia/feed/announcement/AnnouncementClient.kt [hunk 5]
```diff
@@ -85,9 +85,9 @@ class AnnouncementClient : FeedClient {
             if (Prefs.ignoreDateForAnnouncements) {
                 return true
             }
-            return if (announcement.startTime() != null && announcement.startTime().after(date)) {
+            return if (announcement.startTime() != null && announcement.startTime()!!.after(date)) {
```
[the remark this answers is not in the document]
Comment (dbrant) [reply app/src/main/java/org/wikipedia/feed/announcement/AnnouncementClient.kt:88]: yes, in fact I made it even more readable than that.

**Comment by dbrant** [comment]
@sharvaniharan  More good catches!
Should be all fixed now.

**Review by cooltey — APPROVED** [review]

**Comment by sharvaniharan** [comment]
@dbrant there is an error while loading Notifications.. I think you will get a logcat stack trace for it, please let me know if you dont and I will upload a screenshot.

**Comment by dbrant** [comment]
@sharvaniharan  I'm not seeing it; please post the stack trace you see. (it probably has to do with the specific notifications you have for your account)

**Comment by sharvaniharan** [comment]
Screen Shot 2021-09-28 at 10 52 56 AM

**Comment by dbrant** [comment]
@sharvaniharan  ...Fixed!

**Pull request merged**